File: Golang/longestPalindromeSubstr.go
// Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

package main

func longestPalindrom(s string) string {
	start := 0
	end := 0

	for i := 0; i < len(s); i++ {
		l1 := expandAroundCenter(s, i, i)
		l2 := expandAroundCenter(s, i, i+1)

		len := Max(l1, l2)

		if len > (end - start) {
			start = i - (len-1)/2
			end = i + len/2
		}
	}
	return s[start:(end + 1)]
}

func expandAroundCenter(s string, left, right int) int {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left--
		right++
	}
	return right - left - 1
}

// Max return maximum of 2 integer values
func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: Cplusplus/longestPalindromeSubstr.cpp
// Find longeset palindorme substring in given string

#include <iostream>
#include <algorithm>

using namespace std;

string longestPalindrome(string s);
int helper(string s, int left, int right);
bool checkPalindrome(string s);

int main(int argc, char const *argv[])
{
    // string s = "babad";
    // string s = "cbbd";
    string s = "c";

    cout << longestPalindrome(s) << endl;

    return 0;
}

// Taking index as center and expanding it.
string longestPalindrome(string s)
{
    int start = 0;
    int end = 0;
    
    for (int i = 0; i < s.length(); i++)
    {
        int len1 = helper(s, i, i);
        int len2 = helper(s, i, i + 1);

        int len = max(len1, len2);

        if (len > end - start)
        {
            start = i - (len - 1) / 2;
            end = i + len / 2;
        }
    }

    cout << start << ", " << end - start + 1 << endl;
    return s.substr(start, (end - start + 1));
}

int helper(string s, int left, int right)
{
    while (left >= 0 && right < s.length() && s[left] == s[right])
    {
        left--;
        right++;
    }
    return right - left - 1;
}